Decode Unicode in XML requests earlier.

cowl opened the pull request:
Previously, Unicode in input requests was only decoded when parsing the
XML. However, requests are already processed before that.

Unicode is now decoded at an earlier point, so that requests are saved
to the database with proper encoding. This means that the behaviour of
`parseXML` changes: passing improperly encoded Unicode to it will now
result in gibberish.

Relates to the ticket http://ideastest.science.uu.nl/trac/ticket/240

nsbgn committed:
Decode Unicode in XML requests earlier.

Previously, Unicode in input requests was only decoded when parsing the
XML. However, requests are already processed before that.

Unicode is now decoded at an earlier point, so that requests are saved
to the database with proper encoding. This means that the behaviour of
`parseXML` changes: passing improperly encoded Unicode to it will now
result in gibberish.

Relates to the ticket http://ideastest.science.uu.nl/trac/ticket/240

mvangeest commented:
Although our application (https://github.com/UURAGE/ScenarioReasoner/) currently uses an old version of Ideas and I haven't tested this PR myself, I'm fairly certain will affect the handling of JSON requests. This PR decodes the input very early, before a choice is made between `processXML` and `processJSON`. `Ideas.Text.JSON.parseJSON` assumes the input consists of UTF-8 code points: https://github.com/ideas-edu/ideas/blob/0e09adb4a5392315c623806d4ff4587c20c2ba39/src/Ideas/Text/JSON.hs#L166 For this PR, it is probably necessary to remove the `UTF8.decodeM` call from that line, although that breaks the symmetry between `parseJSON` and `prettyJSON`. This would be a breaking, but fixable, change for applications that use either function internally (such as ours).

The encoding detection performed by `Ideas.Text.XML.Unicode.decoding` is based on https://www.w3.org/TR/REC-xml/#charencoding, but it defaults to UTF-8, which is the right choice for JSON too.

It looks like `decoding` also needs to be added to `defaultCommandLine` (otherwise decoding will never happen for command-line "requests").

cowl commented:
Good points. `parseJSON` and `parseXML` should act similarly - and so should `defaultCGI` and `defaultCommandLine`, of course.

I do feel that a decoding step or encoding step doesn't belong in a function that acts on `String`s - but to avoid breaking changes, would it be preferable to make two versions of the `parseXML` function?

mvangeest commented:
>I do feel that a decoding step or encoding step doesn't belong in a function that acts on `String`s

I completely agree. Perhaps now would be a good opportunity to start storing encoded text as `[Word8]` - or the "modern alternative", `ByteString`, which is already the format of incoming data in `defaultCGI` because of [wai](https://hackage.haskell.org/package/wai). A few years ago I mentioned this (in passing) in https://github.com/ideas-edu/ideas/issues/6#issuecomment-222164495. It would be a significant change, but clearly an improvement in type expressiveness, and maybe there are now enough people willing to review it.

>to avoid breaking changes, would it be preferable to make two versions of the `parseXML` function?

As a maintainer of an Ideas application, I'm generally in favour of breaking changes. Releasing a new version of Ideas (fortunately) does not cause older versions to stop working, and I prefer having one "proper" way of doing something in each version. Of course, I cannot speak on behalf of the Ideas maintainer(s).

nsbgn committed:
Add Unicode decoding symmetry.

`parseJSON` now no longer handles decoding, just like `parseXML`.
`defaultCommandLine` now handles decoding, just like `defaultCGI`.

Untested (no test suites)

cowl commented:
I'd be in favour of that too. It's for @BastiaanHeeren to decide - I could take a look if it's desirable. For now, I at least added the essential changes to `parseJSON` and `defaultCommandLine`.
